snakemake/logging.py

```diff
@@ -305,12 +243,504 @@ def log_handler(self, msg):
         self.check_response(response, "/update_workflow_status")
 
 
+class DefaultFormatter(_logging.Formatter):
+
+    def __init__(self, printreason=False, show_failed_logs=False, printshellcmds=False):
+        self.printreason = printreason
+        self.show_failed_logs = show_failed_logs
+        self.printshellcmds = printshellcmds
+        self.last_msg_was_job_info = False
+
+    def format(self, record):
+        """
+        Override format method to handle Snakemake-specific log messages.
+        """
+        msg = record.msg
+        level = msg.get("level", "INFO")  # Default to "INFO" if level not in message
+
+        # Call specific handlers based on the log level
+        if level == "info":
+            return self.handle_info(msg)
+        elif level == "host":
+            return self.handle_host(msg)
+        elif level == "job_info":
+            return self.handle_job_info(msg)
+        elif level == "group_info":
+            return self.handle_group_info(msg)
+        elif level == "job_error":
+            return self.handle_job_error(msg)
+        elif level == "group_error":
+            return self.handle_group_error(msg)
+        elif level == "progress":
+            return self.handle_progress(msg)
+        elif level == "job_finished":
+            return self.handle_job_finished(msg)
+        elif level == "shellcmd":
+            return self.handle_shellcmd(msg)
+        elif level == "rule_info":
+            return self.handle_rule_info(msg)
+        elif level == "d3dag":
+            return self.handle_d3dag(msg)
+        elif level == "dag_debug":
+            return self.handle_dag_debug(msg)
+        elif level == "run_info":
+            return self.handle_run_info(msg)
+        else:
+            return msg["msg"]
```

Comment:
I guess this is something for Python's new match statement.

Reply:
Good idea, agreed.